### ISSUE TITLE
Do releases

### ISSUE THREAD

**Issue by probonopd** [AppImage/AppImageKit#849]
The lack of formal releases is [cited](https://twitter.com/Det_Conan_Kudo/status/1036395262166622208) as a reason for poor support of the AppImage format in distributions (like Fedora), desktops (like GNOME), and IDEs (like GNOME Builder).

> You guys have made far too many choices favoring agility without any security.

Hence, we should do formal releases.

How can we do so in a way that does not create too much additional administrative work for us?

* Do time-based (e.g., monthly) releases
* Fully automate the process
* Use a versioning scheme that doesn't break (or approximates) semver (e.g., `3.0.$DATE`) (1=klik, 2=klik2/cmg, 3=AppImage)

What should we formally release?
* libappimage
* appimagetool
* linuxdeploy(qt)
* libappimageupdate

**Comment by Conan-Kudo** [comment]
I'd prefer to see pretty much everything have formal releases with tarballs and such (especially given the heavy use of submodules). Ideally, the regular releases means that submodules aren't as necessary since the dependencies can be expressed more normally in CMake.

**Comment by probonopd** [comment]
Yes, tarballs are automatically generated by GitHub when one does a tagged release. 

> Ideally, the regular releases means that submodules aren't as necessary since the dependencies can be expressed more normally in CMake.

How would releases enable this? Specifically, say we release once a month. We still want continuous versions of our own dependencies in our continuous builds. It's fine though that others (distributions) would use the monthly releases of all of our stuff, though.

 Would that satisfy your requirements?

**Comment by probonopd** [comment]
Now, what's the easiest way to get Travis CI to do automated time-based releases, say, on the 1st of each month, without any manual intervention? https://docs.travis-ci.com/user/cron-jobs/

**Comment by azubieta** [comment]
IMHO, doing releases is a required thing. But it implies more administrative effort. You will have to maintain at least a master branch (which should be ultra stable) and a development branch (which will be used for our development job and will have continuous builds). Merging into master should never be automated. Bugs can sleep into the stable branch in the final second.

Which tools should get versioned this way? The ones that provide a public API, it doesn't matter if it's a CLI or a C header it would be required in order to inform our users. linuxdeploy will be a special case as it uses the AppDir spec as API with its plugins.

**Comment by TheAssassin** [comment]
@probonopd for linuxdeploy, it's already planned to provide releases in the future. We'll do semantic-like versioning in intervals, e.g., every 1-2 weeks. That way, we have a widely known, comprehensible release scheme, but don't add too much maintenance overhead. Patch versions will _only_ be released to fix critical bugs. which cannot wait until the next release date.

**Comment by Conan-Kudo** [comment]
> You will have to maintain at least a master branch (which should be ultra stable) and a development branch (which will be used for our development job and will have continuous builds).

This is absolutely not required. In nearly all of my projects, I deliberately avoid this workflow because it complicates things too much. Tagging off of the main branch that code is accepted in is perfectly fine, especially since you more or less operate in a fork+pull model.

**Comment by TheAssassin** [comment]
>> You will have to maintain at least a master branch (which should be ultra stable) and a development branch (which will be used for our development job and will have continuous builds).
>
> This is absolutely not required. In nearly all of my projects, I deliberately avoid this workflow because it complicates things too much. Tagging off of the main branch that code is accepted in is perfectly fine, especially since you more or less operate in a fork+pull model.

Exactly. I don't want to do that either. I plan to tag off the master branch, too.

**Comment by probonopd** [comment]
> IMHO, doing releases is a required thing.

Define "required". We did fine for this for the first decade of the project, by keeping all of AppImage in just one repo and doing nothing but continuous releases.

The complexity comes in since we now have a library that outside users might start using.

> But it implies more administrative effort. 

If I don't one thing in my hobby projects, it is administrative effort. In fact if my hobby comes with administrative effort then I will choose other hobbies ;-)

> You will have to maintain at least a master branch (which should be ultra stable) and a development branch (which will be used for our development job and will have continuous builds). 

This is dangerously going into traditional software engineering (great for larger projects and companies where people are paid to do jobs) and exactly the opposite of where I want to go (a hobby project that is fun to hack on in our free time).

I want this to be continuously released, the master branch needs always to be shippable, everyone (users and developers) use the same continuous version, it must never break (and if it does it is rolled back immediately, reverting any questionable commits). I don't want to deal with anything else but the master branch. This means that we need a certain level of test automation.

I also have absolutely no intentions to maintain outdated versions that may end up in Debian stable etc. over time. This is why I am not keen on having this stuff in distributions in the first place. Ideally we would _only_ get this into debian backports (so that debian stable users can get the latest AppImage stuff) but I guess this is not how distros work. Which is why I invented AppImage in the first place ;-)

Merging into master should never be automated. Bugs can sleep into the stable branch in the final second.

**Comment by shoogle** [comment]
> The complexity comes in since we now have a library that outside users might start using.

Sounds like you should start by doing releases only for the library, and only think about doing releases for the other components once you have got the hang of it.

**Comment by Calinou** [comment]
> I don't think automatic tagging is possible

It is possible to create tags on repositories using the GitHub API, see [uploadtool](https://github.com/probonopd/uploadtool) :slightly_smiling_face:

**Comment by shoogle** [comment]
@Calinou, it wasn't the ability to create tags that I doubted, but the ability to do a scheduled build on Travis. However, I now see that [this is also possible](https://docs.travis-ci.com/user/cron-jobs/).

**Comment by TheAssassin** [comment]
Automatic tagging is a horrible idea. Let the developers decide which commit is stable enough to be released. Otherwise, half-finished or buggy work might possibly released.

**Comment by probonopd** [comment]
> Otherwise, half-finished or buggy work might possibly released.

Well... Half-finished work should _ideally_ never be pushed to the master branch. Buggy work should _ideally_ never pass the test automation.

Eveything else means _more admin work for someone_ (not me, this is my hobby, and I don't like to do admin work in my free time).

**Comment by TheAssassin** [comment]
We could even tag using the GitHub web interface.

**Comment by probonopd** [comment]
Looks like we have code in `uploadtool` for this already:
https://github.com/probonopd/uploadtool/blob/master/upload.sh#L32-L37

**Comment by probonopd** [comment]
Started writing a script that automates this but seemingly a Travis job can't push to git by default, what am I missing? Do we need to use the GitHub API and a GitHub token for this? Any volunteers to complete https://github.com/AppImage/libappimage/blob/master/travis/tag-using-cron.sh? (I don't want something that creates _recurring manual work_ but am looking for a "set once and forget" solution.)

**Comment by TheAssassin** [comment]
I'm strongly opposed to auto-tagging. That doesn't make _any_ sense, and is counterproductive.

**Comment by probonopd** [comment]
Then "do releases" means additional admin work for us, which I want to avoid. I am not sold onto the idea anyhow. Our project is all about continuous releases, upstream distros should just treat each git commit to master as a "release".

**Comment by TheAssassin** [comment]
As a maintainer of various software, I have some experience in that topic. Not as much as other people, but I think I can evaluate the costs and benefits of different versioning types. I have done continuous deployment, time-based releases and stable/version branch stuff. My suggestions in that topic are based on that experience.

In my opinion, it will be a _lot_ more work to have to maintain yet another script that somehow is bound to proprietary infrastructure (Travis & GitHub). We have some of these scripts, and they cause trouble all over the place. Error checks are missing or cannot be implemented properly, sometimes one build works and uploads its files, whereas the other one is failing, so architectures are missing, etc. That's quite annoying. The only way to solve these problems is usually to self-host a service that interacts with both other services, allowing to implement all the business logic properly. However, within AppImage, that's usually not a viable strategy.

In fact, we already have all we need to make "timed releases". Just push a tag and Travis will do the rest. For me it's not a problem to add some entry to my calendar to create those tags. Automating this just creates a lot more issues than it solves, human interaction is required for versioning.

Also, releasing can help improving software, as one needs to think about changes more than once. One needs to make sure not to break the own or someone else's software, or if they do, document that properly. That's something we used to lack anyway, but we're improving that recently.
I'm not at all opposed to versioning, as long as it makes sense. For linuxdeploy, users request it, as corporate environments need that for a whole lot of reasons (including internal reviews (security or something), silly politics, etc.). Versioning linuxdeploy in this case is a great idea: we can ship our continuous software under the hood of versioned linuxdeploy releases. For instance, we ship the latest continuous release available of appimagetool in the releases of linuxdeploy, etc.

For our _tools_ I don't want any versioning either. That was one of the main reasons to extract libappimage -- we can version that properly, improve the code, make the API stable, etc., and at the same time keep the continuous release scheme tools using that library (AppImageLauncher, appimagetool, ...).

Your main argument is always "work". Well, of course it's work. Things don't happen magically. But see it like this: you have a team of people _willing_ to _do_ that work, but your own politics sometimes prevent us from bringing AppImage forward. We managed to e.g., get KDE to develop their own AppImage specific technologies by just improving our library and providing it to them separately (which was great anyway, as the code used to be messy and buggy).

@Conan-Kudo is totally right that we already have some "manual" update work in the case of submodules. However, for us this supports our continuous release scheme. We only update dependencies when it's necessary, that way we don't have to continuously update our own software when we introduce breaking changes into internal libraries.
Thinking about versioning in libappimage, there's not much that changes. In fact, we're improving our "bad" behavior when introducing versioning in libappimage in the future. We need to stabilize our APIs. And we just update our submodules to the latest release commits then. Not much difference to now.

**Comment by probonopd** [comment]
Hm. @Conan-Kudo you are making thoughtful points, which I appreciate and why I always like to discuss with you.

> Then you're not actually solving the problem and once again you've proved my point about the double standard.

The AppImage project is about creating a distribution-agnostic agile infrastructure that allows application authors to get out their software to end users without going through the hell that is distribution packaging with all the rules, policies, waiting times, etc.

The AppImage project does this by providing application authors with tools to make AppImages. Since these tools are "applications" too, the same rules (getting out or stuff directly to our users without going through the hell that is distribution packaging with all the rules, policies, waiting times, etc.) should apply to us, too.

> Your tooling is effectively "platform" and "infrastructure". 

No, the target system (=GNU/Linux/Xorg/glib/Gtk+/...) stack is the _platform_. Platform code needs to be stable, make guarantees, and not change (without prior notice). I would not say that our code has reached (or will ever reach) that level.

> That means you have a responsibility to stabilize it and provide predictable and useful releases so that they can be integrated and used to support applications and services.

I am getting your point, and it's not without merits, only that _we_ (or, more precisely, _I_) don't really want to do this. We publish AppImageSpec and ideall the desktop (GNOME, KDE, XFCE,...) people would take and implement these specs to increase usability. Like the people who wrote the ELF spec certainly did not write/maintain all ELF implementation in use today themselves. I have never said we should write _system_infrastructure software_ (as the AppImage project) that other projects should build into their own solutions.

**Comment by probonopd** [comment]
> In fact, we already have all we need to make "timed releases". Just push a tag and Travis will do the rest. For me it's not a problem to add some entry to my calendar to create those tags. 

Exactly.

> Automating this just creates a lot more issues than it solves, human interaction is required for versioning.

Really? Scheduling something in a calendar just to go through a series of manual repetitive steps sounds wasteful to me and it's just not something I enjoy to do in my free time. 

> Also, releasing can help improving software, as one needs to think about changes more than once. One needs to make sure not to break the own or someone else's software, or if they do, document that properly.

That's the bigger point here, imho. We should strive for (I am not saying I am always perfect at it) to only commit stuff that _works_, and if it doesn't, _revert it immediately_. All other work ceases immediately if `master` is broken (does not build or creates showstopper issues when being used). Since we "eat our own dogfood" everywhere (i.e., use continuous builds of our tools ourselves) we are usually the first ones to encounter such issues, which should give us room to fix or rollback.

> I'm not at all opposed to versioning, as long as it makes sense. For linuxdeploy, users request it, as corporate environments need that for a whole lot of reasons (including internal reviews (security or something), silly politics, etc.). 

Couldn't those corporations mirror a certain commit (binary) and have the same advantages? Only that they would have to refer to it as, e.g., 8c8516f, rather than 3.20180806.0.

> For our tools I don't want any versioning either. That was one of the main reasons to extract libappimage -- we can version that properly, improve the code, make the API stable, etc., and at the same time keep the continuous release scheme tools using that library (AppImageLauncher, appimagetool, ...).

Yes, having a thing such as libappimage brings in responsibilities (and expectations) that we didn't have before, when all of our code was only used by our own tools and we were free to change anything at any time. With libappimage, this is no longer the case, as @Conan-Kudo (rightly!) points out. Which is why I am looking for a way to prevent people using libappimage in other way other than linking to it statically, to ensure it is never seen as something stable that can be relied on, but merely as something you have to take care of delivering in the corresponding version yourself if you use it. (I don't want to see or support today's libappimage 5 years from now when it is in Debian oldstable!)

> Your main argument is always "work". Well, of course it's work. Things don't happen magically. 

Don't get me wrong, the whole project is "work", but _creative_, not _administrative_. For me this is purely a hobby, and I simply don't like to do unpaid administrative work.

> But see it like this: you have a team of people willing to do that work

That's great. Just keep in mind:

* If someone volunteers to do a __one-time effort__ (e.g., fix a bug) that's just awesome because it provides value without increasing future burden/effort/debt
* If someone starts work that from thereon needs to be repeated, this is increasing future burden/effort/debt. This is awesome only if the person agrees to carry forward that work indefinitely, but basically no one can do that - hence creating burden/effort/debt for others, which is not that awesome. I have seen this multiple times: Someone moves, gets married, starts to work for Google, etc.; priorities shift and the project suffers

So, in summary: If someone is stepping up to do the work, I am the last person to stop them. Just think about the implications, and don't expect _me_ to do repetitive administrative work.

**Comment by TheAssassin** [comment]
Yes, in the end it's a lot less work, as fixing scripts and issues created by scripts is something super annoying (we have these bugs every couple of weeks in AppImage, and they take me hours to analyze and fix usually...). Clicking a button in a project after thinking about it for 2 minutes whether it's a good idea to release that commit is a lot less time intensive.

> don't expect me to do repetitive administrative work.

Nobody does. And nobody guarantees the development of libappimage etc. anyway. If someone doesn't maintain the software any more, there is no point in creating releases any more.

**Comment by probonopd** [comment]
> If someone doesn't maintain the software any more, there is no point in creating releases any more.

You have a point there 👍

**Comment by develar** [comment]
At least, thanks for Release build (11) :) It crashes for me, but at least now I am aware that electron-builder should update AppImage runtime because now it is clear that currently used is outdated. As was already stated, without formal releases I don't know should I update or not. And without changelogs not clear for me for what I should spend my time and update. But changelogs it is another topic.

**Comment by probonopd** [comment]
> At least, thanks for Release build (11) :) It crashes for me

Hi @develar, have you opened a ticket for it?

> As was already stated, without formal releases I don't know should I update or not.

Since electron-builder is only using the runtime, I expect there won't be frequent changes, and generally the latest continuous build should be the best one, as we don't do maintenance branches.

**Comment by probonopd** [comment]
If it's no work, then I ask someone to do it :+1:

**Comment by TheAssassin** [comment]
Nobody else expected this to be real work (https://github.com/AppImage/AppImageKit/issues/849#issuecomment-419732854).
